### ISSUE TITLE
Add post-AccumulateGrad hook as a nice public API

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#76464]
This is used by DDP to trigger synchronization of the gradients and we also have other users that want the optimizer to run alongside the backward pass.

Today a workaround is doing `t.weight.view_as(t.weight).grad_fn.next_functions[0][0].register_hook(my_hook)` but that is pretty shady.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by soulitzer** [comment]
> if we call register_hook directed on parameters, will it work?

Calling register hooks on parameters registers a tensor prehook, which would fire BEFORE the gradients are actually accumulated.

**Comment by albanD** [comment]
Will let @janeyx99 comment on this as she's working on this exact feature.

**Comment by janeyx99** [comment]
@sunpengsdu Like Jeffrey said, using register_hook will not work. However! I am working on a PR that will introduce a way to register the type of hook you'd like (tentatively called register_post_accumulate_grad_hook in the PR), and it will be easier and less shady to have the optimizer along in the backward pass. See https://github.com/pytorch/pytorch/pull/107063

**Comment by sunpengsdu** [comment]
Thanks a lot

**Comment by gau-nernst** [comment]
Hello, is there (or plan to have) a hook for post-accumulate-grad at `nn.Module` level (instead of at `Tensor` level)? From #86051, it indicates that `nn.Module.register_full_backward_hook()` does not guarantee `grad_weight` is ready when the hook fires. It would be useful for something I'm trying to implement. Thank you.

**Comment by soulitzer** [comment]
@gau-nernst  Registering a [register_multi_grad_hook](https://pytorch.org/docs/stable/autograd.html#torch.autograd.graph.register_multi_grad_hook) to the parameters (+ input) of a module sounds similar to the API that you are describing. Note though that such a hook would fire when the gradients are ready, but potentially prior to .grad being populated. Is that okay for your case?

**Comment by gau-nernst** [comment]
@soulitzer Thank you for your suggestion, it sounds great! I will try it out.

If it works well, perhaps I will raise a feature request (in a separate GitHub issue) to add it `nn.Module` for convenience? It can be simple as `register_multi_grad_hook(list(self.parameters()), hook)`
